### PR TITLE
Add login nodes configuration parameters

### PR DESCRIPTION
**Pull request by hehe7318**
### Description of changes
* Create classes to cluster_config.py with full configuration parameters without update policy:
* Create above classes’ schema to cluster_schema.py with validators
* Create LoginNodeIam class in cluster_config.py and for better maintainability
        * Create _BaseIam class with the fields in common between all the Iams
        * Modify Iam class to become HeadNodeAndQueueIam inherited from _BaseIam class
        * Make LoginNodeIam class inherited from _BaseIam class
* Centralise AMI id Regexp in [constants](https://github.com/aws/aws-parallelcluster/blob/develop/cli/src/pcluster/constants.py) file for better maintainability
* Create BaseIam related classes in cluster_config.py and for better maintainability
        * Make LoginNodesIam and HeadNodeIam inherit from BaseIam
* Modify DummySlurmClusterConfig to add support for login nodes
* Unit test
    * Test cluster config
        * test login node pool default value
    * Test cluster schema validators
        * Add unit test to make sure the custom ami is validate
        * Add unit test to make sure login node pool count must be greater than zero
        * Add unit test to make sure login node pool is validate
        * Add unit test to confirm for the MVP, only 1 pool can be under the LoginNodes section
        * Add unit test to confirm that the instance profile and instance role cannot be used together.
    * Modify yaml file to add support for login nodes

### Sample YAML configuration
```
LoginNodes:
  Pools: 
  - Name: test
    InstanceType: t2.micro
    Count: 1
    Image:
        CustomAmi: ami-12345678
    Networking:
      SubnetId: subnet-12345678901234567
    Ssh:
      KeyName: test-us-east-2
    Iam:
      InstanceRole: arn:aws:iam::aws:role/LoginNodeRole
```

### Tests

* Unit test run locally

### Checklist

- [x] Make sure you are pointing to the right branch and add a label in the PR title (i.e. 2.x vs 3.x)
- [x] Check all commits' messages are clear, describing what and why vs how.
- [x] Make sure to have added unit tests or integration tests to cover the new/modified code.
- [x] Check if documentation is impacted by this change.

### CODE REVIEW

File: cli/src/pcluster/schemas/cluster_schema.py [hunk 1]
```diff
@@ -1257,6 +1262,66 @@ def make_resource(self, data, **kwargs):
         return HeadNode(**data)
 
 
+class LoginNodeImageSchema(BaseSchema):
+    custom_ami = fields.Str(validate=validate.Regexp(r"^ami-[0-9a-z]{8}$|^ami-[0-9a-z]{17}$"))
```
Comment (EddyMM) [inline cli/src/pcluster/schemas/cluster_schema.py]: The regular expression used to check if an AMI id is valid seems to be similar for the `HeadNodeImageSchema` and `QueueImageSchema`.
Let's centralise it in a [constants](https://github.com/aws/aws-parallelcluster/blob/develop/cli/src/pcluster/constants.py) file for better maintainability.

File: cli/src/pcluster/schemas/cluster_schema.py [hunk 2]
```diff
@@ -1257,6 +1262,66 @@ def make_resource(self, data, **kwargs):
...
+    @post_load
+    def make_resource(self, data, **kwargs):
+        """Generate resource."""
+        return LoginNodeSsh(**data)
+
+
+class LoginNodeNetworkingSchema(BaseNetworkingSchema):
+    subnet_id = fields.Str(required=True)
+    elastic_ip = fields.Str()
+    proxy = fields.Str()
```
Comment (EddyMM) [inline cli/src/pcluster/schemas/cluster_schema.py]: I don't think these parameters are being used (for now).

File: cli/src/pcluster/config/cluster_config.py [hunk 3]
```diff
@@ -1183,6 +1195,102 @@ def __init__(
...
+
+
+class LoginNodeNetworking(_BaseNetworking):
+    """Represent the networking configuration for the login node."""
+
+    def __init__(
+            self,
+            subnet_id: str,
+            security_groups: List[str] = None,
+            additional_security_groups: List[str] = None,
```
Comment (EddyMM) [inline cli/src/pcluster/config/cluster_config.py]: I think the `security_groups` and `additional_security_groups` arguments can be consideed to be part of the `kwargs` argument.
We can then let the base class [_BaseNetworking](https://github.com/aws/aws-parallelcluster/blob/develop/cli/src/pcluster/config/cluster_config.py#L655) read them.

File: cli/src/pcluster/validators/cluster_validators.py [hunk 4]
```diff
@@ -1470,3 +1470,22 @@ def _validate(self, queues):
                     f"public IPs can only be assigned to instances launched with a single network interface.",
                     FailureLevel.ERROR,
                 )
+
+
+class PoolsValidator(Validator):
```
Comment (EddyMM) [inline cli/src/pcluster/validators/cluster_validators.py]: This can be verified at Schema level using the [marshmallow Length validator](https://marshmallow.readthedocs.io/en/stable/marshmallow.validate.html#marshmallow.validate.Length)

File: cli/src/pcluster/config/cluster_config.py [hunk 5]
```diff
@@ -1183,6 +1195,102 @@ def __init__(
...
+                         **kwargs)
+        self.subnet_id = Resource.init_param(subnet_id)
+
+    def _register_validators(self, context: ValidatorContext = None):
+        super()._register_validators(context)
+        self._register_validator(AvailabilityZoneValidator, login_node_subnet_id=self.subnet_id,
+                                 head_node_subnet_id=HeadNodeNetworking.getInstance().subnet_id)
+
+    @property
+    def availability_zone(self):
```
Comment (EddyMM) [inline cli/src/pcluster/config/cluster_config.py]: Where is this property used? If it's unused then we can remove it.

File: cli/src/pcluster/schemas/cluster_schema.py [hunk 6]
```diff
@@ -1257,6 +1262,66 @@ def make_resource(self, data, **kwargs):
...
+
+    @post_load
+    def make_resource(self, data, **kwargs):
+        """Generate resource."""
+        return LoginNodeNetworking(**data)
+
+
+class LoginNodePoolSchema(BaseSchema):
+    name = fields.Str(required=True, validate=[validate.Length(min=1), validate.Regexp(r'^[a-z0-9]+$')])
```
Comment (EddyMM) [inline cli/src/pcluster/schemas/cluster_schema.py]: The Length=1 validation should probably be in the `LoginNodesSchema`.
Also, let's remove this regexp for now (to be consistent with other naming checks).

File: cli/src/pcluster/schemas/cluster_schema.py [hunk 7]
```diff
@@ -1257,6 +1262,66 @@ def make_resource(self, data, **kwargs):
...
+    instance_type = fields.Str(required=True)
+    image = fields.Nested(LoginNodeImageSchema)
+    networking = fields.Nested(LoginNodeNetworkingSchema, required=True)
+    count = fields.Int(required=True, validate=validate.Range(min=1))
+    ssh = fields.Nested(LoginNodeSshSchema, required=True)
+    iam = fields.Nested(IamSchema)
+    gracetime_period = fields.Int(validate=validate.Range(max=120))
+    admin_user = fields.Str(required=True)
```
Comment (EddyMM) [inline cli/src/pcluster/schemas/cluster_schema.py]: Let's remove these parameters for now (let's sync on this).

File: cli/src/pcluster/config/cluster_config.py [hunk 8]
```diff
@@ -1183,6 +1195,102 @@ def __init__(
...
+        super().__init__(**kwargs)
+        self.name = Resource.init_param(name)
+        self.instance_type = Resource.init_param(instance_type)
+        self.image = image
+        self.networking = networking
+        self.count = Resource.init_param(count, default=1)
+        self.ssh = ssh
+        self.iam = iam or Iam(implied=True)
+        self.gracetime_period = Resource.init_param(gracetime_period, default=60)
+        self.admin_user = Resource.init_param(admin_user, default=getpass.getuser())
```
Comment (EddyMM) [inline cli/src/pcluster/config/cluster_config.py]: Let's remove these parameters for now.

**Review by EddyMM — CHANGES_REQUESTED** [review]
Good job on your first PR in this repo!

I've left a few comments. Plus there are some failing unit tests that need to be fixed e.g. [CLI tests](https://github.com/aws/aws-parallelcluster/actions/runs/5097687429/jobs/9164363957?pr=5358)

File: cli/src/pcluster/config/cluster_config.py (hunk 8 above)
Comment (hehe7318) [reply cli/src/pcluster/config/cluster_config.py]: Done! I have removed them.

File: cli/src/pcluster/schemas/cluster_schema.py (hunk 7 above)
Comment (hehe7318) [reply cli/src/pcluster/schemas/cluster_schema.py]: Done! I have removed them.

File: cli/src/pcluster/schemas/cluster_schema.py (hunk 6 above)
Comment (hehe7318) [reply cli/src/pcluster/schemas/cluster_schema.py]: I added Length(min=1) validation because I thought the length of the 'name' string should be longer than 0. But I have removed it. And also removed the regexp.

File: cli/src/pcluster/config/cluster_config.py (hunk 5 above)
Comment (hehe7318) [reply cli/src/pcluster/config/cluster_config.py]: It was originally intended for availability_zone validator, but I modified the validator implementation so it's not needed now, and I forgot to remove it. Now removed.

File: cli/src/pcluster/validators/cluster_validators.py (hunk 4 above)
Comment (hehe7318) [reply cli/src/pcluster/validators/cluster_validators.py]: Done. I have moved it to Schema level.

File: cli/src/pcluster/config/cluster_config.py (hunk 3 above)
Comment (hehe7318) [reply cli/src/pcluster/config/cluster_config.py]: That make sense. My bad! I have modified it.

File: cli/src/pcluster/schemas/cluster_schema.py (hunk 2 above)
Comment (hehe7318) [reply cli/src/pcluster/schemas/cluster_schema.py]: I forgot to remove them. Now they were removed.

File: cli/src/pcluster/schemas/cluster_schema.py (hunk 1 above)
Comment (hehe7318) [reply cli/src/pcluster/schemas/cluster_schema.py]: Done!

File: cli/src/pcluster/config/cluster_config.py [hunk 9]
```diff
@@ -1308,6 +1416,7 @@ def __init__(
         head_node: HeadNode,
         scheduling=None,
         shared_storage: List[Resource] = None,
+        login_nodes: LoginNodes = None,
```
Comment (EddyMM) [inline cli/src/pcluster/config/cluster_config.py]: LoginNodes should be a section associated with the `SlurmClusterConfig` instead of `BaseClusterConfig` for now (to avoid it being included in the `AWSBatchConfig`).

File: cli/src/pcluster/config/cluster_config.py [hunk 10]
```diff
@@ -678,6 +684,12 @@ def availability_zone(self):
         """Compute availability zone from subnet id."""
         return AWSApi.instance().ec2.get_subnet_avail_zone(self.subnet_id)
 
+    @staticmethod
+    def getInstance():
```
Comment (EddyMM) [inline cli/src/pcluster/config/cluster_config.py]: Is this logic included in order to get the head node's subnet ID from the `LoginNodeNetworking` resource?
If so, I would recommended registering the validator at a higher layer in the configuration heirarchy (In this case, the `SlurmClusterConfig` resource). That way you can access both subnets using the

**Review by EddyMM — CHANGES_REQUESTED** [review]

File: cli/src/pcluster/config/cluster_config.py (hunk 10 above)
Comment (hehe7318) [reply cli/src/pcluster/config/cluster_config.py]: Done! It's really a good idea!

File: cli/src/pcluster/config/cluster_config.py (hunk 9 above)
Comment (hehe7318) [reply cli/src/pcluster/config/cluster_config.py]: Got blockers. Can we sync on this?

Comment (hehe7318) [reply cli/src/pcluster/config/cluster_config.py]: Done!

File: cli/src/pcluster/config/cluster_config.py [hunk 11]
```diff
@@ -668,6 +669,7 @@ def __init__(self, subnet_id: str, elastic_ip: Union[str, bool] = None, proxy: P
         self.subnet_id = Resource.init_param(subnet_id)
         self.elastic_ip = Resource.init_param(elastic_ip)
         self.proxy = proxy
+        HeadNodeNetworking._instance = self
```
Comment (EddyMM) [inline cli/src/pcluster/config/cluster_config.py]: I don't think this is being used.

File: cli/tests/pcluster/templates/test_cluster_stack.py [hunk 12]
```diff
@@ -15,6 +15,7 @@
 from abc import ABC, abstractmethod
 from datetime import datetime
 
+import pyperclip
```
Comment (EddyMM) [inline cli/tests/pcluster/templates/test_cluster_stack.py]: This won't be needed in the test.

File: cli/src/pcluster/config/cluster_config.py [hunk 13]
```diff
@@ -971,6 +970,27 @@ def _register_validators(self, context: ValidatorContext = None):  # noqa: D102
...
+        self,
+        s3_access: List[S3Access] = None,
+        **kwargs,
+    ):
+        super().__init__(**kwargs)
+        self.s3_access = s3_access
+
+
+class LoginNodeIam(_BaseIam):
+    """Represent the IAM configuration for HeadNode and Queue."""
```
Comment (EddyMM) [inline cli/src/pcluster/config/cluster_config.py]: `... IAM configuration for login node.`

File: cli/src/pcluster/config/cluster_config.py [hunk 14]
```diff
@@ -1183,6 +1203,79 @@ def __init__(
...
+    def __init__(self, custom_ami: str, **kwargs):
+        super().__init__()
+        self.custom_ami = Resource.init_param(custom_ami)
+
+    def _register_validators(self, context: ValidatorContext = None):  # noqa: D102 #pylint: disable=unused-argument
+        if self.custom_ami:
+            self._register_validator(CustomAmiTagValidator, custom_ami=self.custom_ami)
+
+
+class LoginNodeSsh(Ssh):
```
Comment (EddyMM) [inline cli/src/pcluster/config/cluster_config.py]: Let's handle this like the IAM scenario (i.e. Inheritance) since the `Ssh` section for `LoginNodes` doesn't have an Allowed IPs parameter for now.

The Allowed IPs parameter might be an implied value if we keep it as it is now but the schema should try to match the resource to ensure it's not unintentionally used downstream.

File: cli/src/pcluster/config/cluster_config.py (hunk 13 above)
Comment (EddyMM) [reply cli/src/pcluster/config/cluster_config.py]: Add/Extend a unit test to confirm that the instance profile and instance role cannot be used together.

File: cli/src/pcluster/config/cluster_config.py [hunk 15]
```diff
@@ -971,6 +970,27 @@ def _register_validators(self, context: ValidatorContext = None):  # noqa: D102
             self._register_validator(InstanceProfileValidator, instance_profile_arn=self.instance_profile)
 
 
+class HeadNodeAndQueueIam(_BaseIam):
```
Comment (EddyMM) [inline cli/src/pcluster/config/cluster_config.py]: We can simply call it `Iam` instead of `HeadNodeAndQueueIam` to avoid confusing the name with what it does. Essentially, it's used in either the HeadNode or the Queue but the term `HeadNodeAndQueueIam` might give the impression that the IAM values apply to both the head node and compute nodes at the same time (which is not the case).

File: cli/src/pcluster/schemas/cluster_schema.py [hunk 16]
```diff
@@ -928,6 +931,14 @@ def no_coexist_role_policies(self, data, **kwargs):
                 "InstanceProfile, InstanceRole or AdditionalIamPolicies can not be configured together."
             )
 
+
+class HeadNodeAndQueueIamSchema(BaseIamSchema):
```
Comment (EddyMM) [inline cli/src/pcluster/schemas/cluster_schema.py]: Same suggestion, use `IamSchema`.

File: cli/src/pcluster/schemas/cluster_schema.py [hunk 17]
```diff
@@ -1257,6 +1282,64 @@ def make_resource(self, data, **kwargs):
...
+        """Generate resource."""
+        return LoginNodePool(**data)
+
+
+class LoginNodesSchema(BaseSchema):
+    pools = fields.Nested(
+        LoginNodePoolSchema,
+        many=True,
+        required=True,
+        validate=validate.Length(min=1, max=1, error="has a minimum size of 1, "
```
Comment (EddyMM) [inline cli/src/pcluster/schemas/cluster_schema.py]: The error message will probably be revised later but for now, let's go with a tentative message: "Only one pool can be specified when using login nodes."

Comment (EddyMM) [reply cli/src/pcluster/schemas/cluster_schema.py]: Also, you can use equal instead of setting a min and max of 1 ([Length](https://marshmallow.readthedocs.io/en/stable/marshmallow.validate.html#marshmallow.validate.Length))

File: cli/src/pcluster/validators/networking_validators.py [hunk 18]
```diff
@@ -198,3 +198,14 @@ def _validate_all_subnets_exist(self, existing_subnets, expected_subnet_ids):
...
             )
+
+
+class AvailabilityZoneValidator(Validator):
+
+    def _validate(self, login_node_subnet_id, head_node_subnet_id):
+        if AWSApi.instance().ec2.get_subnet_avail_zone(login_node_subnet_id) != \
+                AWSApi.instance().ec2.get_subnet_avail_zone(head_node_subnet_id):
+            self._add_failure(
+                "LoginNode Networking SubnetId must be in the same availability zone as the HeadNode.",
```
Comment (EddyMM) [inline cli/src/pcluster/validators/networking_validators.py]: Messaging will need rewording but let's go for a tentative message: "Login nodes and head node must be in the same availability zone".

File: cli/tests/pcluster/config/test_cluster_config.py [hunk 19]
```diff
@@ -1,4 +1,5 @@
 import pytest
+import getpass
```
[the remark this answers is not in the document]
Comment (EddyMM) [reply cli/tests/pcluster/config/test_cluster_config.py]: ^^^

**Review by EddyMM — CHANGES_REQUESTED** [review]

File: cli/src/pcluster/validators/networking_validators.py (hunk 18 above)
Comment (hehe7318) [reply cli/src/pcluster/validators/networking_validators.py]: Done!

File: cli/src/pcluster/schemas/cluster_schema.py (hunk 17 above)
Comment (hehe7318) [reply cli/src/pcluster/schemas/cluster_schema.py]: Done!

File: cli/src/pcluster/schemas/cluster_schema.py (hunk 16 above)
Comment (hehe7318) [reply cli/src/pcluster/schemas/cluster_schema.py]: Done!

File: cli/src/pcluster/config/cluster_config.py (hunk 15 above)
Comment (hehe7318) [reply cli/src/pcluster/config/cluster_config.py]: Done!

File: cli/src/pcluster/config/cluster_config.py (hunk 11 above)
Comment (hehe7318) [reply cli/src/pcluster/config/cluster_config.py]: I forgot to remove it. Done!

File: cli/tests/pcluster/config/test_cluster_config.py (hunk 19 above)
[the remark this answers is not in the document]
Comment (hehe7318) [reply cli/tests/pcluster/config/test_cluster_config.py]: Done!

File: cli/tests/pcluster/templates/test_cluster_stack.py (hunk 12 above)
Comment (hehe7318) [reply cli/tests/pcluster/templates/test_cluster_stack.py]: Forgot to remove it. Done

File: cli/src/pcluster/config/cluster_config.py (hunk 14 above)
Comment (hehe7318) [reply cli/src/pcluster/config/cluster_config.py]: Done!

File: cli/src/pcluster/config/cluster_config.py (hunk 13 above)
Comment (hehe7318) [reply cli/src/pcluster/config/cluster_config.py]: Done!

**Review by EddyMM — APPROVED** [review]
